lib/os/topological.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+#include <stdlib.h>
+#include <zephyr/sys/slist.h>
+
+LOG_MODULE_REGISTER(topo, LOG_LEVEL_INF);
+
+struct topo_edge {
+	int sink_pos;
+	sys_snode_t node;
+};
+
+struct topo_point {
+	uint16_t data;
+	sys_slist_t edge_list;
+};
+
+struct topo_graph {
+	int point_num;
+	int edge_num;
+	struct topo_point *points;
+};
+
+static inline int get_vertex_position(struct topo_graph *graph, uint16_t data)
+{
+	int i;
+
+	__ASSERT(graph, "topological graph struct should not be NULL");
+
+	for (i = 0; i < graph->point_num; i++)
+		if (graph->points[i].data == data)
+			return i;
+
+	return -1;
+}
+
+static int sorting(struct topo_graph *graph, uint16_t *order)
+{
+	int i, j, index = 0;
+	struct topo_edge *edge;
+	int *queue, *in_degree;
+	int head = 0, tail = 0;
+	int ret = 0;
+
+	__ASSERT(graph, "topological graph struct should not be NULL");
+	__ASSERT(order, "order array should not be NULL");
+
+	queue = (int *)malloc(graph->point_num * sizeof(int));
```

Comment:
Use of libc malloc limits the application here to things that can reasonably run in the generic heap.   Drivers and subsystems very often have their own memory; consider adding the ability to use a specified k_heap (where e.g. a NULL pointer means "use the system heap")

Reply:
use k_heap_alloc() instead of malloc()